src/ProgressLogging.jl

```diff
@@ -3,13 +3,168 @@ module ProgressLogging
 # Use README as the docstring of the module:
 @doc read(joinpath(dirname(@__DIR__), "README.md"), String) ProgressLogging
 
-export @progress, @withprogress, @logprogress
+export @progress, @progressid, @withprogress, @logprogress
 
 using Base.Meta: isexpr
+using UUIDs: UUID, uuid4, uuid5
 using Logging: Logging, @logmsg, LogLevel
 
 const ProgressLevel = LogLevel(-1)
 
+"""
+    ProgressLogging.ROOTID
+
+This is used as `parentid` of root [`Progress`](@ref)es.
+"""
+const ROOTID = UUID(0)
+
+"""
+    ProgressLogging.Progress(id, [fraction]; [parentid, name, done])
+
+# Usage: Progress log record provider
+
+Progress log record can be created by using the following pattern
+
+```julia
+id = uuid4()
+try
+    @info Progress(id)  # create a progress bar
+    # some time consuming task
+    # ...
+    @info Progress(id, 0.1)  # update progress to 10%
+    # ...
+finally
+    @info Progress(id, done = true)  # close the progress bar
+end
+```
+
+It is recommended to use [`@withprogress`](@ref),
+[`@logprogress`](@ref), and optionally [`@progressid`](@ref) to create
+log records.
+
+# Usage: Progress log record consumer (aka progress monitor)
+
+It is recommended to use [`ProgressLogging.asprogress`](@ref) instead
+of checking `message isa Progress`.  Progress monitors can retrieve
+progress-related information from the following properties.
+
+# Properties
+- `fraction::Float64`: it can take following values:
+  - `0 <= fraction < 1`
+  - `fraction = NaN`: indeterminate progress
+  - `fraction >= 1`: completed
+- `id::UUID`: Identifier of the task whose progress is at `fraction`.
+- `parentid::UUID`: The ID of the parent progress.  It is set to
+  [`ProgressLogging.ROOTID`](@ref) when there is no parent progress.
+  This is used for representing progresses of nested tasks.  Note that
+  sub-tasks may be executed concurrently; i.e., there can be multiple
+  child tasks for one parent task.
+- `name::String`: Name of the progress bar.
+- `done::Bool`: `true` if the task is done.
+"""
+Base.@kwdef struct Progress
+    id::UUID
+    parentid::UUID = ROOTID  # not nested by default
+    fraction::Float64 = NaN
```

Comment:
Is it better to use `fraction::Union{Missing,Float64} = missing` here?

Reply:
Yes, I think.

Reply:
`nothing` might also be ok in this context? I think `missing` implies there should be some value but it was unavailable. `nothing` that there's definitively no value (cf. `findfirst`).

Either way, the `Union` is a better way to go I think.

Reply:
Right, it's also close to the old/untyped API which allowed `progress = nothing`:

https://github.com/JunoLab/ProgressLogging.jl/blob/0eae8c65950be40ceb7e20900c58bc66f3bd3024/src/ProgressLogging.jl#L20-L23

I think now I'm inclined to use `fraction::Union{Nothing,Float64} = nothing`.